src/http/server.cr

```diff
@@ -465,6 +468,7 @@ class HTTP::Server
           if io
             # a non nillable version of the closured io
             _io = io
+            @client_connection_count.add 1
```

Comment:
It could be an option to move this call to `handle_client`, so it's placed in the same scope as the corresponding decrease.
To avoid any sync issues, we can make sure the newly spawned fiber executes directly:
```cr
client_fiber = spawn handle_client(io)
Fiber.current.enqueue
client_fiber.resume
```

I'm not sure which effect this would have on performance. Time to first byte could be improved, though.

An alternative could be to refactor this method in a way that once accepted, handling the client connection happens in the same fiber, and a new fiber is spawned for the next `accept` call.
Example:
```cr
def listen
  @socket.each do |socket|
    spawn accept_socket(socket)
  end
end
private def accept_socket(socket)
  loop do
    io = begin
      socket.accept?
    rescue e
      handle_exception(e)
      next
    end

    if io
      # a non nillable version of the closured io
      _io = io
      spawn accept_socket(socket)
      handle_client(_io)
    else
      break
    end
  end
end
```

Reply:
> WDYT about keeping a reference to each active client connection? That would allow doing a hard(er) shutdown by closing them from the server when close times out.

What should the behavior be?
*  `SHUT_RD` then wait timeout seconds for sending to finish.  websockets?.  When a timeout is reached should it close remaining sockets or let the fibers linger?  I'm not sure there's harm in letting the fibers linger as some may complete before process exit or whatever else the program does.  As long as remaining fibers don't hang or display errors exiting the program
* `SHUT_RDWR` aborts everything.  I think this defeats the purpose of graceful shutdown
* Custom handlers that allow signaling of shutdown.   HTTP 1.x defaults to `SHUT_RD`.  This could go in a later PR improving websockets.  I don't use websockets so it's not my concern but if anyone else does please respond

Reply:
I think there only needs to be a single timeout which allows for graceful shutfown. If that times out, there's nothing else left but doing a hard shutdown. The current behaviour, just letting ˋ´#listenˋ return is IMO not really useful.